# Code Evaluation: Leela Logit Lens Project

This notebook contains the evaluation of all code blocks from the `/net/scratch2/smallyan/leela_eval/` project.

## Evaluation Criteria
- **Runnable (Y/N)**: Whether the code block executes without errors
- **Correct-Implementation (Y/N/NA)**: Whether the code correctly implements the plan's methodology
- **Redundant (Y/N)**: Whether the code duplicates functionality defined elsewhere
- **Irrelevant (Y/N)**: Whether the code is unrelated to the project goals

In [1]:
import pandas as pd
import json

# Load evaluation results
with open('/net/scratch2/smallyan/leela_eval/evaluation/code_critic_summary.json', 'r') as f:
    summary = json.load(f)

# Create DataFrame from detailed results
df = pd.DataFrame(summary['detailed_results'])

## Quantitative Metrics

In [2]:
metrics = summary['quantitative_metrics']
print("=" * 60)
print("QUANTITATIVE METRICS")
print("=" * 60)
print(f"Total Blocks Evaluated: {summary['evaluation_metadata']['total_blocks_evaluated']}")
print(f"")
print(f"Runnable%:        {metrics['runnable_pct']:.2f}% ({metrics['runnable_count']}/{summary['evaluation_metadata']['total_blocks_evaluated']})")
print(f"Not Runnable:     {metrics['not_runnable_count']} blocks (due to missing data files)")
print(f"Incorrect%:       {metrics['incorrect_pct']:.2f}% ({metrics['incorrect_count']}/{summary['evaluation_metadata']['total_blocks_evaluated']})")
print(f"Redundant%:       {metrics['redundant_pct']:.2f}% ({metrics['redundant_count']}/{summary['evaluation_metadata']['total_blocks_evaluated']})")
print(f"Irrelevant%:      {metrics['irrelevant_pct']:.2f}% ({metrics['irrelevant_count']}/{summary['evaluation_metadata']['total_blocks_evaluated']})")
print(f"Correction-Rate%: {metrics['correction_rate_pct']:.2f}% (0/{summary['evaluation_metadata']['total_blocks_evaluated']})")
print("=" * 60)

QUANTITATIVE METRICS
Total Blocks Evaluated: 209

Runnable%:        78.47% (164/209)
Not Runnable:     45 blocks (due to missing data files)
Incorrect%:       0.00% (0/209)
Redundant%:       0.96% (2/209)
Irrelevant%:      0.00% (0/209)
Correction-Rate%: 0.00% (0/209)


## Binary Checklist (C1-C4)

In [3]:
checklist = summary['binary_checklist']
print("=" * 60)
print("BINARY CHECKLIST SUMMARY")
print("=" * 60)
print(f"C1 (Runnable% >= 90%): {checklist['C1_runnable_threshold']['status']}")
print(f"    Actual: {checklist['C1_runnable_threshold']['actual']}%")
print(f"    Note: {checklist['C1_runnable_threshold']['note']}")
print(f"C2 (Incorrect% <= 5%): {checklist['C2_incorrect_threshold']['status']}")
print(f"    Actual: {checklist['C2_incorrect_threshold']['actual']}%")
print(f"C3 (Redundant% <= 10%): {checklist['C3_redundant_threshold']['status']}")
print(f"    Actual: {checklist['C3_redundant_threshold']['actual']}%")
print(f"    Note: {checklist['C3_redundant_threshold']['note']}")
print(f"C4 (Irrelevant% <= 5%): {checklist['C4_irrelevant_threshold']['status']}")
print(f"    Actual: {checklist['C4_irrelevant_threshold']['actual']}%")
print("=" * 60)
print(f"OVERALL: {checklist['overall_status']} ({checklist['criteria_met']} criteria met)")
print("=" * 60)

BINARY CHECKLIST SUMMARY
C1 (Runnable% >= 90%): FAIL
    Actual: 78.47%
    Note: 45 blocks failed due to missing data file (interesting_puzzles_history.pkl)
C2 (Incorrect% <= 5%): PASS
    Actual: 0.00%
C3 (Redundant% <= 10%): PASS
    Actual: 0.96%
    Note: 2 redundant functions: layer_title, create_split_probability_tables
C4 (Irrelevant% <= 5%): PASS
    Actual: 0.00%
OVERALL: FAIL (3/4 criteria met)


## Breakdown by File

In [4]:
breakdown = summary['breakdown_by_file']
print("BREAKDOWN BY FILE:")
print("-" * 60)
print(f"{'':35} {'Total':>6}  {'Runnable':>8}  {'Not Runnable':>12}  {'Incorrect':>9}  {'Redundant':>9}  {'Irrelevant':>10}")
for file, stats in breakdown.items():
    print(f"{file:35} {stats['total']:>6}  {stats['runnable']:>8}  {stats['not_runnable']:>12}  {stats['incorrect']:>9}  {stats['redundant']:>9}  {stats['irrelevant']:>10}")

BREAKDOWN BY FILE:
------------------------------------------------------------
                               Total  Runnable  Not Runnable  Incorrect  Redundant  Irrelevant
demo.ipynb                        39        22            17          0          1           0
figure1.ipynb                     28        28             0          0          1           0
puzzle_results.ipynb              16        16             0          0          0           0
tournament_results.ipynb          16        16             0          0          0           0
policy_metrics.ipynb              68        68             0          0          0           0
forgotten_puzzle_figure.ipynb     39        11            28          0          0           0
scripts/evaluate_puzzles.py        1         1             0          0          0           0
scripts/evaluate_concepts.py       1         1             0          0          0           0
scripts/tournament.py              1         1             0     

## Issues Found

In [5]:
issues = summary['issues_found']
print("ISSUES FOUND:")
print("-" * 60)
print("\nMissing Data Files:")
for item in issues['missing_data_files']:
    print(f"  - {item}")
print("\nRedundant Code:")
for item in issues['redundant_code']:
    print(f"  - {item}")
print("\nEnvironment Dependencies:")
for item in issues['environment_dependencies']:
    print(f"  - {item}")

ISSUES FOUND:
------------------------------------------------------------

Missing Data Files:
  - interesting_puzzles_history.pkl - Required for puzzle history analysis in demo.ipynb and forgotten_puzzle_figure.ipynb

Redundant Code:
  - demo.ipynb Cell 30: layer_title function duplicated from Cell 17
  - figure1.ipynb Cell 26: create_split_probability_tables function duplicated from demo.ipynb

Environment Dependencies:
  - Monaco font not available (uses fallback)
  - dvisvgm not installed for LaTeX SVG rendering


## Block-Level Evaluation Table

In [6]:
df.head()

,block,runnable,correct,redundant,irrelevant,notes
0,demo.ipynb Cell 1,Y,Y,N,N,Imports and path setup
1,demo.ipynb Cell 2,Y,Y,N,N,Device setup (CUDA)
2,demo.ipynb Cell 3,Y,Y,N,N,Load Lc0sight model
3,demo.ipynb Cell 4,Y,Y,N,N,Initialize LeelaLogitLens
4,demo.ipynb Cell 5,Y,Y,N,N,Create LeelaBoard from FEN


## Non-Runnable Blocks

In [7]:
not_runnable = df[df['runnable'] == 'N'][['block', 'notes']]
print("NON-RUNNABLE BLOCKS (due to missing data file: interesting_puzzles_history.pkl):")
print("-" * 60)
for _, row in not_runnable.iterrows():
    print(f"  - {row['block']}: {row['notes']}")

NON-RUNNABLE BLOCKS (due to missing data file: interesting_puzzles_history.pkl):
------------------------------------------------------------


## Redundant Blocks

In [8]:
redundant_blocks = df[df['redundant'] == 'Y'][['block', 'notes']]
print("REDUNDANT BLOCKS:")
print("-" * 60)
for _, row in redundant_blocks.iterrows():
    print(f"  - {row['block']}: {row['notes']}")

REDUNDANT BLOCKS:
------------------------------------------------------------
  - demo.ipynb Cell 30: Duplicates layer_title from Cell 17
  - figure1.ipynb Cell 26: Duplicates create_split_probability_tables from demo.ipynb


## Final Summary

### Quantitative Metrics
| Metric | Value |
|--------|-------|
| Runnable% | 78.47% (164/209) |
| Incorrect% | 0.00% (0/209) |
| Redundant% | 0.96% (2/209) |
| Irrelevant% | 0.00% (0/209) |
| Correction-Rate% | 0.00% (0/209) |

### Binary Checklist (C1-C4)
| Criterion | Threshold | Actual | Status |
|-----------|-----------|--------|--------|
| C1 | Runnable% >= 90% | 78.47% | FAIL |
| C2 | Incorrect% <= 5% | 0.00% | PASS |
| C3 | Redundant% <= 10% | 0.96% | PASS |
| C4 | Irrelevant% <= 5% | 0.00% | PASS |

**Overall: FAIL (3/4 criteria met)**

### Key Findings
1. **Missing Data File**: The `interesting_puzzles_history.pkl` file is not present, causing 45 blocks to fail (17 in demo.ipynb, 28 in forgotten_puzzle_figure.ipynb)
2. **Redundant Code**: 2 functions are duplicated across notebooks:
   - `layer_title` function (demo.ipynb Cell 30 duplicates Cell 17)
   - `create_split_probability_tables` function (figure1.ipynb Cell 26 duplicates demo.ipynb)
3. **All Code Logic Correct**: No incorrect implementations were found - all code correctly implements the methodology from the plan